File: src/main/scala/assignment1/Group.scala
package assignment1

class Group[T]:
  private var members = List[T]()

  def add(value: T): Unit = {
    if (!has(value)) {
      members = value :: members
    }
  }

  def delete(value: T): Unit = {
    members = members.filterNot(_ == value)
  }

  def has(value: T): Boolean = {
    members.contains(value)
  }

object Group:
  def from[T](iterable: Iterable[T]): Group[T] = {
    val group = new Group[T]()
    for (value <- iterable) {
      group.add(value)
    }
    group
  }


@main def test() = {
  val group = Group.from(List(10, 20))
  println(group.has(10))
  // → true
  println(group.has(30))
  // → false
  group.add(10)
  group.delete(10)
  println(group.has(10))
  // → false
}

File: src/main/js/assignment1/Group.js
class Group {
    #contents;

    constructor() {
        this.members = [];
    }

    has(value) {
        return this.members.includes(value);
    }

    add(value) {
        if (!this.has(value)) {
          this.members.push(value);
        }
    }

    delete(value) {
        this.members = this.members.filter(member => member !== value);
    }

    static from(iterable) {
        let group = new Group();
        for (let value of iterable) {
          group.add(value);
        }
        return group;
    }
}

function main() {
    let group = Group.from([10, 20]);
    console.log(group.has(10));
    // → true
    console.log(group.has(30));
    // → false
    group.add(10);
    group.delete(10);
    console.log(group.has(10));
    // → false
}

if (require.main === module) {
    main();
}

module.exports = Group;